Graphics MVP

belfortep opened the pull request:
MVP of conway's game of life, with graphics support. How to use the program is explained in the README.md

belfortep committed:
can show conways game on screen

belfortep committed:
added game view constants

belfortep committed:
supports command line arguments and files for points

belfortep committed:
Merge branch 'development' of github.com:belfortep/conways-game-in-rust into graphics

belfortep committed:
Can run game with a makefile

belfortep committed:
cargo fmt and cargo clippy pass

lazcanoluca commented on src/conways/conways_game_view.rs:
```diff
@@ -0,0 +1,55 @@
+                draw_rectangle(x_position, y_position, width, height, BLACK);
+            });
+
+            sleep(Duration::from_secs_f32(CONSTANT_WAIT));
```
Instead of using a `sleep`, which blocks the execution thread, I would consider using a timeout. At low [`CONSTANT_WAIT`](https://github.com/belfortep/conways-game-in-rust/blob/79e214451dede6d50798231592ea29896499efe2/src/conways/conways_game_constants.rs#L5) it works fine, but I tried increasing the constant to slow the simulation down, and then interacting with the window controls (resizing, fullscreen, closing), which was buggy (I'm guessing, because of the execution blocking at the `sleep`).

belfortep replied:
Okay, just searched and there are some function for controlling the time in macroquad, I'm trying with those

lazcanoluca replied:
Nice! I was using `Instant` haha. If there are time utilities I'll feel dumb 🤣

lazcanoluca commented on src/conways/point.rs:
```diff
@@ -29,3 +29,28 @@ impl Point {
     }
 }
+
+impl From<&str> for Point {
+    fn from(s: &str) -> Self {
+        let mut positions = s.split(',');
+
+        let (Some(x_position), Some(y_position), None) =
+            (positions.next(), positions.next(), positions.next())
+        else {
+            return Self::default();
+        };
+
+        let Ok(x_position) = x_position.parse() else {
+            return Self::default();
+        };
+
+        let Ok(y_position) = y_position.parse() else {
+            return Self::default();
+        };
+
+        Self {
+            x_position,
+            y_position,
+        }
+    }
+}
```
I believe it would be better to implement [`TryFrom`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) instead, because the conversion `str` to `Point` can fail if the string is not in the expected format.

belfortep replied:
That's true, i'm going to add some tests also

lazcanoluca replied:
Nice!

belfortep committed:
Changed way of creating points from &str
Now we use TryFrom instead of From, so now when
the string is in wrong format, it returns
a result with an Error

belfortep committed:
changed sleep in update view to timeout

belfortep committed:
pass cargo clippy

lazcanoluca commented on src/conways/conways_game_view.rs:
```diff
@@ -0,0 +1,59 @@
+
+
```
```suggestion
```

lazcanoluca commented on src/conways/conways_game_view.rs:
```diff
@@ -0,0 +1,59 @@
+            if current_frame_time - last_frame_time >= CONSTANT_WAIT {
+                clear_background(WHITE);
+
+            self.conways_game.next_generation();
+
+            self.conways_game.cells_do(|cell| {
+                let width = CELLS_WIDTH;
+                let height = CELLS_HEIGHT;
+                let scale_factor = VIEW_SCALE_FACTOR;
+                let x_position = Self::get_x_position(cell.x_position, scale_factor);
+                let y_position = Self::get_y_position(cell.y_position, scale_factor);
+                draw_rectangle(x_position, y_position, width, height, BLACK);
+            });
```
Broken indentation

lazcanoluca commented:
I left a couple of style related remarks. That stuff should be automagically be taken care of by activating the format on save (I think it runs `rust fmt`) option on your IDE. Otherwise, lgtm 🚀 🚀 🚀

belfortep committed:
cargo fmt and cargo clippy pass
